compiler-core/src/type_/expression.rs

```diff
@@ -1046,6 +1046,57 @@ impl<'a, 'b> ExprTyper<'a, 'b> {
                 }
             }
 
+            ClauseGuard::FieldAccess {
+                location,
+                label,
+                container,
+                ..
+            } => {
+                let typed_container = self.infer_clause_guard(*container.clone())?;
+
+                match typed_container.type_().as_ref() {
+                    Type::App { .. } => match typed_container.clone() {
```

Comment:
It doesn't look like we do anything with the knowledge that this is a `Type::App` here, so this flow control can be removed.